### ISSUE TITLE
Not working with socket.io-redis-emitter

### ISSUE THREAD

**Issue by dzeroone** [socketio/socket.io-redis-streams-adapter#8]
I have tried to emit from a cronjob, but it's not reaching the client!

**Comment by darrachequesne** [comment]
Yes, the Redis emitter is meant to be used with the classic Redis adapter, we will have to implement a custom emitter publishing to the Redis streams.

**Comment by fjeldstad** [comment]
> Yes, the Redis emitter is meant to be used with the classic Redis adapter, we will have to implement a custom emitter publishing to the Redis streams.

Is this something that is planned?

**Comment by fjeldstad** [comment]
Are there any significant drawbacks of just creating a dummy socket.io server with the `redis-streams-adapter` configured and using `emit` on that server instance to send events from background processes (like the cron job mentioned in the original post)? I tested this locally and as far as I can see it works well; no need to call `listen(...)` or anything, just create a server instance and emit 🤷

**Comment by gmarcos87** [comment]
Hi @darrachequesne, is there any alternative to use the adapter/emitter pattern with this stream library? I would like to be able to have `connection status retrieval` and also a emitter for other micro-services (or event sources).

**Comment by eggsys** [comment]
is this mean if i want to create custom Emitter like other serivice without socket.io dependency
i can direct Emit payload into Redis Stream using redis-streams-adapter ?
@darrachequesne 



> Are there any significant drawbacks of just creating a dummy socket.io server with the `redis-streams-adapter` configured and using `emit` on that server instance to send events from background processes (like the cron job mentioned in the original post)?
> 
> I've tested this locally both with the old `redis-adapter` and this `redis-streams-adapter` and as far as I can see it works well; no need to call `listen(...)` or anything, just create a server instance and emit 🤷 Seems too good to be true! Is it?

**Comment by eggsys** [comment]
> is this mean if i want to create custom Emitter like other serivice without socket.io dependency i can direct Emit payload into Redis Stream using redis-streams-adapter ? [@darrachequesne](https://github.com/darrachequesne)
> 
> > Are there any significant drawbacks of just creating a dummy socket.io server with the `redis-streams-adapter` configured and using `emit` on that server instance to send events from background processes (like the cron job mentioned in the original post)?
> > I've tested this locally both with the old `redis-adapter` and this `redis-streams-adapter` and as far as I can see it works well; no need to call `listen(...)` or anything, just create a server instance and emit 🤷 Seems too good to be true! Is it?


i achieve to Emit message into redis Stream by using 
`RedisStreamsAdapter` class
with `.doPublish`


its will look like this

```
export class Emitter {
    private adapter: RedisStreamsAdapter;
    private nsp!: any;



    constructor(redisClient: any, opts?: any) {
    const options = Object.assign(
    {
      streamName: "socket.io",
      maxLen: 200,
      readCount: 100,
      sessionKeyPrefix: "sio:session:",
      heartbeatInterval: 5_000,
      heartbeatTimeout: 10_000,
    },
    opts
  );

    console.log("Initializing Emitter with options:", this.nsp);
    this.nsp = this.getNsp();
    this.adapter = new RedisStreamsAdapter(this.nsp, redisClient, options);
    console.log("Emitter initialized with RedisStreamsAdapter");
    }

    public getAdapter(): RedisStreamsAdapter {
        return this.adapter;
    }

    public getNsp(): any {


        const fakeNsp = {
            name: "chat",
            server: {
                opts: {},
                encoder: new Encoder()
            }
        };

        return fakeNsp;
    }

    public async publish(message?: ClusterMessage): Promise<void> {
        const payload: ClusterMessage = {
          uid: "xxxxxxxx"
          nsp: "/chat",
          type: 3,
          data: {
            packet: {
              type: 2,
              data: [
                "chat",
                {
                  userId: "xxxxxxxxxx",
                  message: "Message from Emitter",
                },
              ],
            },
            opts: {
              rooms: ["general-chat"],
              except: [],
              flags: {},
            },
          },
        };

        // uid: xxxxxxx, nsp: /chat, type: 2
        console.log("Publishing message:", payload);


        const payload2 :ClusterMessage = {
            uid: "xxxxxxxxx",
            nsp: "/chat",
            type: 2,
        }
        await this.adapter.doPublish(payload);
        await this.adapter.doPublish(payload2);

    }
}
```

**Comment by fjeldstad** [comment]
What I have done - for now - is to create a Socket.io server with the `@socket.io/redis-streams-adapter` and then use that to emit events. It does not listen to any port (so it does not have `engine` set). Important to remember to call `close()` on the server on disposal, otherwise it tends to block tests from exiting etc. - but note that calling that function _will_ throw an error due to `engine` not being set. It does properly dispose of the adapter though.

Example:

```typescript
import { createAdapter } from '@socket.io/redis-streams-adapter';
import { Server } from 'socket.io';
import { Redis } from 'ioredis';

// Create a Socket.io server using the Redis streams adapter
const redis = new Redis();
const io = new Server({
  adapter: createAdapter(redis)
});

// Emit events
io.to('some-socket-or-room').emit('test-event', { /* ... */ });

// Disposal
try {
  await io.close(); // This will throw a type error (after closing the adapter)
} catch (error) {
  // Ignore the expected error
}
redis.disconnect();

```

It would be really nice @darrachequesne if someone from the team could share your thoughts about this approach, and also if you are planning to develop a more light-weight emitter package for the Redis streams adapter.

**Comment by darrachequesne** [comment]
Here you go:

```js
import { createClient } from "redis";
import { Emitter } from "@socket.io/redis-streams-emitter";

const redisClient = createClient({
  url: "redis://localhost:6379"
});

await redisClient.connect();

const io = new Emitter(redisClient);

setInterval(() => {
  io.emit("ping", new Date());
}, 1000);
```

Source: https://github.com/socketio/socket.io/tree/main/packages/socket.io-redis-streams-emitter
NPM: https://www.npmjs.com/package/@socket.io/redis-streams-emitter
